### ISSUE TITLE
Posts with global post type restriction not excluded from search queries

### ISSUE THREAD

**Issue by nosegraze** [restrictcontentpro/restrict-content-pro#1689]
(when `Hide Restricted Posts` is checked on)

We redirect away from post type archives, but don't do anything to remove restricted post types from other page types.

Report: https://secure.helpscout.net/conversation/529442842/76795?folderId=841776

**Comment by tnorthcutt** [comment]
They are also not hidden from the main blog archive, category archives, date archives, etc.

Interestingly, posts that had individual restrictions set before an entire post type was restricted _are_ hidden from those archives, even though those per-post restrictions should have no effect once an entire post type is restricted.

**Comment by tnorthcutt** [comment]
I made an initial pass at this, with mixed results:

```php
$restricted_post_types    = array_keys( rcp_get_restricted_post_types() );

if ( ! empty( $restricted_post_types ) ) {
	$all_types = array_keys( get_post_types() );
	$remaining = array_diff( $all_types, $restricted_post_types );
	$query->set( 'post_type', $remaining );
}
```

The logic here is that if any post types are restricted, we need to remove them from the query. Since `WP_Query` doesn't have an e.g. `post_type_not_in` var, instead we fetch all post types, then remove our restricted post types.

But, in some initial testing, this produces undesirable results, because then some other post types get included that wouldn't normally be. For instance, on a site with the homepage set to display latest posts, and the `post` post type restricted, pages get shown.

Not sure of the best solution here.

**Comment by mindctrl** [comment]
Not sure I love this, but thinking out loud here...

We may need to save a list of post IDs for restricted post types and use those to set `post__not_in`, but that's a list that has to be managed on the various post action hooks (add/delete/update) and also we'd have to listen for post type change events, since you're technically able to change the post type of a post - we could compare existing post type to submitted post type in the `wp_insert_post_data` filter.

**Comment by tnorthcutt** [comment]
I see where you're going with that... not trying to sound harsh at all, but I'm also not sure I love it 😉. Just sounds rough trying to keep that list up to date.

I wonder how expensive it would be to do the following:
1. Check for restricted post types
2. Do a `WP_Query` for all post types in the resulting array
3. Pass the resulting array of post IDs to `post__not_in`

@mindctrl what do you think?

**Comment by tnorthcutt** [comment]
@mindctrl I went ahead with that method; see https://github.com/restrictcontentpro/restrict-content-pro/pull/1704

**Comment by nosegraze** [comment]
Honestly I don't love this method of getting all the post IDs of restricted post types. Our current way of getting the IDs of all restricted posts is already problematic (see https://github.com/restrictcontentpro/restrict-content-pro/issues/135#issuecomment-286413905 ). This may fix the query issue, but it's ultimately making an existing problem even worse by introducing more possibly intensive queries to get a bunch more post IDs.

For now, I'd prefer a modification of @tnorthcutt 's original suggestion here https://github.com/restrictcontentpro/restrict-content-pro/issues/1689#issuecomment-370085170 . Instead of getting all post types on the site, we can just get the value that's already set in the current query: `$query->get( 'post_type' )`, then remove all restricted post types from there and re-set the new value. This should prevent irrelevant post types from being included when they shouldn't be.

I'd rather we wait until we've fixed the intensive query issue (by moving to taxonomies or whatever) before we try including post type restricted posts in `rcp_get_restricted_post_ids()`.

**Comment by tnorthcutt** [comment]
@nosegraze in my testing, the query doesn't have a `post_type` var set by default, which is why I abandoned that method 😕

Not sure if there's a way around that.

**Comment by nosegraze** [comment]
@tnorthcutt What template were on on when you tested that?

**Comment by tnorthcutt** [comment]
`index.php` in the twentysixteen theme (on the home page, with the home page set to display latest posts).

**Comment by nosegraze** [comment]
In that case it's possible `post_type` was blank because the default value is just `post`.

**Comment by tnorthcutt** [comment]
Yep, agreed. Do you know of a way around that, then?

**Comment by nosegraze** [comment]
I would then just assume the value is `post` if the `post_type` parameter is empty. And if the whole `post` post type is restricted then we should be handling it via the redirect here https://github.com/restrictcontentpro/restrict-content-pro/issues/1686#issuecomment-368758229

Do you see any issue with that?

**Comment by tnorthcutt** [comment]
@nosegraze Good point, that makes sense.

Sounds like `post_type` is set on the query, and it's a string, then we'll need to compare it against `rcp_get_restricted_post_types()` and possibly unset it. And if `post_type` is set and it's an array, then we'll need to `array_diff` against `rcp_get_restricted_post_types()`.

But if `post_type` is not set, and the `post` post type is restricted, we'll need to do what, exactly? Should we unset that query var, or perform a redirect at that point?

In my testing, with `hide_premium` set, a page chosen for `redirect_from_premium`, and the `post` post type restricted, a non-logged in visitor is not redirected away from the blog archive. Should they be, currently?

Sorry – I'm not clear if you're saying here:

> And if the whole post post type is restricted then we should be handling it via the redirect here #1686 (comment)

that we are currently handling it, or that we _should_ be but aren't at the moment 😄

**Comment by nosegraze** [comment]
> But if post_type is not set, and the post post type is restricted, we'll need to do what, exactly? Should we unset that query var, or perform a redirect at that point?

> Sorry – I'm not clear if you're saying here:
> 
> > And if the whole post post type is restricted then we should be handling it via the redirect here #1686 (comment)
 >
> that we are currently handling it, or that we should be but aren't at the moment 😄

If only one post type is set and that post type is restricted, we should redirect away from the page. This is kind of what we tried to set up but it needs some tweaking. See this comment: https://github.com/restrictcontentpro/restrict-content-pro/issues/1686#issuecomment-368758229

> In my testing, with hide_premium set, a page chosen for redirect_from_premium, and the post post type restricted, a non-logged in visitor is not redirected away from the blog archive. Should they be, currently?

We're _not_ handling that now, but we _should_ be. We only check for `is_post_type_archive()` in the redirects and that returns `false` for the blog archive so it's not actually being covered.

--

So to recap/clarify, here's what I think should be happening:

- If _multiple_ post types are set, simply unset any restricted post types from the query/array. If all post types are restricted, see blow.
- If _one_ post type is set and it's a restricted post type the current user cannot access (or if all post types are restricted), don't bother messing with the query. Instead we should redirect to the specified page.

**Comment by tnorthcutt** [comment]
@nosegraze Thanks for the clarification! https://github.com/restrictcontentpro/restrict-content-pro/pull/1704 is now updated accordingly.
